goalc/compiler/compilation/Asm.cpp

```diff
@@ -429,50 +544,203 @@ Val* Compiler::compile_asm_mul_vf(const goos::Object& form, const goos::Object&
 }
 
-Val* Compiler::compile_asm_mulx_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
+Val* Compiler::compile_asm_mul_x_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
   return compile_asm_vf_math3(form, rest, IR_VFMath3Asm::Kind::MUL,
                               emitter::Register::VF_ELEMENT::X, env);
 }
 
-Val* Compiler::compile_asm_muly_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
+Val* Compiler::compile_asm_mul_y_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
   return compile_asm_vf_math3(form, rest, IR_VFMath3Asm::Kind::MUL,
                               emitter::Register::VF_ELEMENT::Y, env);
 }
 
-Val* Compiler::compile_asm_mulz_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
+Val* Compiler::compile_asm_mul_z_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
   return compile_asm_vf_math3(form, rest, IR_VFMath3Asm::Kind::MUL,
                               emitter::Register::VF_ELEMENT::Z, env);
 }
 
-Val* Compiler::compile_asm_mulw_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
+Val* Compiler::compile_asm_mul_w_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
   return compile_asm_vf_math3(form, rest, IR_VFMath3Asm::Kind::MUL,
                               emitter::Register::VF_ELEMENT::W, env);
 }
 
-Val* Compiler::compile_asm_abs_vf(const goos::Object& form, const goos::Object& rest, Env* env) {
+Val* Compiler::compile_asm_vf_math4_two_operation(const goos::Object& form,
+                                                  const goos::Object& rest,
+                                                  IR_VFMath3Asm::Kind first_op_kind,
+                                                  IR_VFMath3Asm::Kind second_op_kind,
+                                                  emitter::Register::VF_ELEMENT broadcastElement,
+                                                  Env* env) {
   auto args = get_va(form, rest);
   va_check(
-      form, args, {{}, {}},
+      form, args, {{}, {}, {}, {}},
       {{"color", {false, goos::ObjectType::SYMBOL}}, {"mask", {false, goos::ObjectType::INTEGER}}});
   bool color = true;
   if (args.has_named("color")) {
     color = get_true_or_false(form, args.named.at("color"));
   }
 
   auto dest = compile_error_guard(args.unnamed.at(0), env)->to_reg(env);
-  if (!dest->settable() || dest->ireg().reg_class != RegClass::VECTOR_FLOAT) {
-    throw_compiler_error(
-        form, "Invalid destination register for a vector float 3-arg math form. Got a {}.",
-        dest->print());
+  auto src1 = compile_error_guard(args.unnamed.at(1), env)->to_reg(env);
+  auto src2 = compile_error_guard(args.unnamed.at(2), env)->to_reg(env);
+  // This third register is intended for the ACC/Q/ETC, and is used to temporarily store the value
+  // that eventually goes into the destination
+  //
+  // For example VMADD:
```

Comment:
With the `+=`, this would be `VMADDA`, not `VMADD`, right?

I'm guessing we can use this form for both versions, and for `VMADDA`, we can have the decompiler add a `(mov acc vf_dest)`

Reply:
Yeah for my example's purposes you're right, its a `VMADDA`, must have got confused as like you said, we can use it for both.